apps/session-manager/src/utils/lollipop.ts

```diff
@@ -0,0 +1,306 @@
+  ResponseErrorValidation,
+} from "@pagopa/ts-commons/lib/responses";
+import { NextFunction, Request, Response } from "express";
+import { FiscalCode, NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+import { ulid } from "ulid";
+import { Errors } from "io-ts";
+import {
+  LollipopLocalsType,
+  LollipopRequiredHeaders,
+  Thumbprint,
+} from "../types/lollipop";
+import { LollipopApiClient } from "../repositories/lollipop-api";
+import { LollipopSignatureInput } from "../generated/fast-login-api/LollipopSignatureInput";
+import { AssertionRef } from "../generated/lollipop-api/AssertionRef";
+import {
+  JwkPubKeyHashAlgorithm,
+  JwkPubKeyHashAlgorithmEnum,
+} from "../generated/lollipop-api/JwkPubKeyHashAlgorithm";
+import { AssertionRefSha256 } from "../generated/lollipop-api/AssertionRefSha256";
+import { AssertionRefSha384 } from "../generated/lollipop-api/AssertionRefSha384";
+import { AssertionRefSha512 } from "../generated/lollipop-api/AssertionRefSha512";
+import { RedisSessionStorageService } from "../services";
+import { RedisClientSelectorType } from "../types/redis";
+import { generateLCParams } from "../services/lollipop";
+import { DomainErrorTypes } from "../models/domain-errors";
+import { ResLocals } from "./express";
+import { withValidatedOrValidationError } from "./responses";
+import { withOptionalUserFromRequest } from "./user";
+import { log } from "./logger";
+
+const NONCE_REGEX = new RegExp(';?nonce="([^"]+)";?');
+// Take the first occurrence of the field keyid into the signature-params
+const KEY_ID_REGEX = new RegExp(';?keyid="([^"]+)";?');
+
+const getNonceOrUlid = (
+  lollipopSignatureInput: LollipopSignatureInput,
+): NonEmptyString => {
+  // The nonce value must be the first regex group
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [_, nonce, ...__] = NONCE_REGEX.exec(lollipopSignatureInput) ?? [
+    null,
+    ulid(),
+  ];
+  return nonce as NonEmptyString;
+};
+
+export const algoToAssertionRefSet = new Set([
+  { algo: JwkPubKeyHashAlgorithmEnum.sha256, type: AssertionRefSha256 },
+  { algo: JwkPubKeyHashAlgorithmEnum.sha384, type: AssertionRefSha384 },
+  { algo: JwkPubKeyHashAlgorithmEnum.sha512, type: AssertionRefSha512 },
+]);
+
+const getAlgoFromAssertionRef = (
+  assertionRef: AssertionRef,
+): JwkPubKeyHashAlgorithm =>
+  pipe(
+    Array.from(algoToAssertionRefSet),
+    (ar) => ar.find((entry) => entry.type.is(assertionRef)),
+    O.fromNullable,
+    O.map((pubKeyHashAlgo) => pubKeyHashAlgo.algo),
+    O.getOrElseW(() => void 0 as never),
+  );
+
+const getKeyThumbprintFromSignature = (
+  lollipopSignatureInput: LollipopSignatureInput,
+): E.Either<Errors, Thumbprint> => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [_, thumbprint, ...__] = KEY_ID_REGEX.exec(lollipopSignatureInput) ?? [
+    null,
+    null,
+  ];
+  return Thumbprint.decode(thumbprint);
+};
+
+const getAndValidateAssertionRefForUser = (
```

Comment:
ditto

Reply:
refactor present in https://github.com/pagopa/io-auth-n-identity-domain/pull/22/commits/0aac1a7af5e32060657c60e53ef458e41e4e445c